llmfoundry/utils/data_prep_utils.py

```diff
@@ -0,0 +1,144 @@
+# Copyright 2022 MosaicML LLM Foundry authors
+# SPDX-License-Identifier: Apache-2.0
+
+import json
+import os
+import platform
+from glob import glob
+from typing import List, Optional
+
+import psutil
+from composer.utils import ObjectStore
+from torch.utils.data import DataLoader, Dataset
+
+
+def build_dataloader(dataset: Dataset,
+                     batch_size: int,
+                     num_workers: Optional[int] = None) -> DataLoader:
+    if num_workers is None:
+        # Multiple workers is only supported on linux machines
+        if 'linux' in platform.platform().lower():
+            num_workers = max(1, psutil.cpu_count())  # type: ignore
+        else:
+            num_workers = 0
+
+    # If using multiple workers, configure each worker to prefetch as many samples as it can, up to
+    # the aggregate device batch size
+    # If not using workers, the torch DataLoader expects the default value for prefetch_factor,
+    # which non-intuitively must be 2.
+    # If on macOS, PyTorch requires prefetch_factor set to None since num_workers is always zero
+    if 'macos' in platform.platform().lower() or num_workers == 0:
+        prefetch_factor = None
+    else:
+        prefetch_factor = max(1, 2 * batch_size //
+                              num_workers) if num_workers > 0 else 2
+
+    return DataLoader(
+        dataset=dataset,
+        sampler=None,
+        batch_size=batch_size,
+        num_workers=num_workers,
+        prefetch_factor=prefetch_factor,
+    )
+
+
+def with_id(basename: str, shard_id: int) -> str:
+    """Get a new basename with the given shard_id.
+
+    From https://github.com/mosaicml/streaming/blob/main/examples/multiprocess_dataset_conversion.ipynb.
+
+    Args:
+        basename (str): Old basename of file.
+        shard_id (int): New shard ID.
+
+    Returns:
+        str: New basename of file.
+    """
+    parts = basename.split('.')
+    parts[1] = f'{shard_id:05}'
+    return '.'.join(parts)
+
+
+def merge_shard_groups(root: str) -> None:
```

Comment:
This method removes the sub-directory folder, rename the shard files, and put everything in a root folder. This method would work, but I recommend a below method which actually create a new `index.json` in a root directory and the `basename` field in `index.json` points to the shard filepath of a sub-directory. Also, with this method, you won't need `with_id()` method.

```
def merge_index(root: str) -> None:
    """Create a new index.json in a root directory which points to the sub-directory shard files.

    Args:
        root (str): Root directory.
    """
    shard_id = 0
    infos = []
    for sub_dir in os.listdir(out_root):
        index_filename = os.path.join(root, sub_dir, 'index.json')
        obj = json.load(open(index_filename))
        for info in obj['shards']:
            old_basename = info['raw_data']['basename']
            new_basename = os.path.join(sub_dir, old_basename)
            info['raw_data']['basename'] = new_basename

            if info['zip_data'] is not None:
                old_basename = info['zip_data']['basename']
                new_basename = os.path.join(sub_dir, old_basename)
                info['zip_data']['basename'] = new_basename

            shard_id += 1
            infos.append(info)

    index_filename = os.path.join(root, 'index.json')
    obj = {
        'version': 2,
        'shards': infos,
    }
    text = json.dumps(obj, sort_keys=True)
    with open(index_filename, 'w') as out:
        out.write(text)
```

We will update the jupyter notebook with the above method as well. 

Please verify it by running your script end-to-end.

Reply:
@karan6181 Okay if I do this in a follow up refactor? I can do that immediately after getting this in, but this change requires a few changes across files to handle paths differently since originally we assume that the output directory is flat.

Reply:
Sure works for me. Thank You!